### ISSUE TITLE
[Track 4] 압축 규칙 관련 추가 문의

### ISSUE THREAD

**Issue by 2KangHo** [AI-GrandChallenge/round-1#57]
#42 의 3,4에 대한 질문입니다. 답변이 없어 새로 이슈를 답니다.
uint8이나 int8등 8bit로 저장해서 load시 float32로 변환하는 것도 허용하는 건가요??

_Originally posted by @2KangHo in https://github.com/AI-GrandChallenge/round-1/issues/42#issuecomment-664742471_

**Comment by bluebrush** [comment]
@2KangHo 님, 말씀하신 방법은 압축과 유사하여 안되는 것으로 안내해드립니다.

**Comment by 2KangHo** [comment]
@bluebrush 그렇다는건 모델 코드 자체가 int8으로 구성된 경우만 가능하다는 말씀이신가요?
추가로 저장과 load 모두 int8로 하고 inference에서 float32로 바꾸는것도 불허하는 건가요?

**Comment by ClovaAIAdmin** [comment]
네 infer 때는  activation 고려 float32 계산 가능한 것으로 하겠습니다.

**Comment by 2KangHo** [comment]
@ClovaAIAdmin inference시에 weight들이 32bit인 경우면 load시 32bit로 변환하는 것과 차이가 없지 않나요?

**Comment by ClovaAIAdmin** [comment]
저장시 8bit quant 하면서 일부 정보들이 유실될텐데요(물론 태스크에 따라 마지널 합니다). activation과 BN까지 모두 8bit로 막으면 현실적으로 quant는 쓰지마라가 되기 때문입니다. 그래서 저장과 로드는 상징적 의미라고 보시면 되겠습니다.

**Comment by 2KangHo** [comment]
@ClovaAIAdmin 그렇다면 처음 질문이었던 uint8이나 int8등 8bit로 저장해서 load시 float32로 변환하는 것도 허용된다는 부분 아닌가요?

**Comment by ClovaAIAdmin** [comment]
네 내부적으로 논의 해보니 activation이나 BN등 문제로 성능이 너무 저하될 우려가 있어 다른 방법들과 차이가 너무 날 수 있다는 의견이 있어서 load때 type cast는 가능할 것 같습니다. 물론 속도는 현저히 느려지겠지요.

**Comment by JuYong-Shin** [comment]
위 내용을 보고 zip이나 다른 압축 방식을 사용하지 않고, float32를  int8로 바꾸었을 때, 정보 손실을 감수하고, weight file의 크기가  15MB 이만이면 되는 것으로 이해 했습니다. 그러면 모델의 크기는 제한이 없어도 되지 않나요.  torch에서 제공하는 module이 model을 int8로 바꾸는 것을 허용하지 않아서 문제가 발생했습니다. cuda nvcc로 프로그램 작성하면 문제가 되지 않는 상황일 것 같은데, 허용하지 않았었고요.  도움 요청 부탁 드립니다.

**Comment by ClovaAIAdmin** [comment]
모델 크기의 제한을 두는 이유는 경량화 관점에서 아무리 커도 이정도를 넘지마라는 관점입니다. 말씀 주신 모델크기가 파라미터 개수인지 웨이트파일 크기인지 애매한데 웨이트 파일 크기는 제한을 반드시 해야하고요(경량화 과제니까). 파라미터 크기는 웨이트파일크기가 넘지않도록 유동적으로 하시면됩니다. torch의 int8이 문제라면 pytorch 제공 QAT나 저희 statassist (Clova ai github) 참조 하시면 힌트를 얻으실수 있을 것 같습니다. 그리고 경량화 방법이 quantization만 있는 것은 아니어서요.

**Comment by JuYong-Shin** [comment]
int8로 weight 파일을 저장했으므로,  parameter 크기나, 파일 크기가 같습니다. 저희 statassist (Clova ai github) 참조  이부분에 자세히 알려 주시면 감사하겠습니다. 문제는 model을 만들었을 때, nn  module이 int type을 허용하지 않아서, 초기에 사용하지 않지만 모델 크기가 크게 잡히기 때문입니다.  그리고, 다른  경량화도 적용  하려고 하는데,  마찬가지 문제가 발생할 것 같습니다.

**Comment by JuYong-Shin** [comment]
weight 파일 크기는 15MB보다 작습니다.

**Comment by JuYong-Shin** [comment]
statassist (Clova ai github) 참조부분 링크 추가 부탁드립니다.

**Comment by ClovaAIAdmin** [comment]
https://github.com/clovaai/StatAssist-GradBoost

**Comment by JuYong-Shin** [comment]
감사합니다.

**Comment by wysung1** [comment]
요즘 모델 압축 논문들은 웨이트 액티베이션 다 2비트 정도만 사용하는데 모두 8비트로 막으면 안된다니 참으로 이해하기 힘든 말입니다.
